pkg/cmd/server/origin/master_config.go

```diff
@@ -309,6 +285,200 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
+		kubeAdmissionPluginConfigFilename = options.KubernetesMasterConfig.APIServerArguments["admission-control-config-file"][0]
+		hasSeparateKubeAdmissionChain = true
+	}
+
+	openshiftAdmissionPlugins := openshiftAdmissionControlPlugins
+	if len(options.AdmissionConfig.PluginOrderOverride) > 0 {
+		openshiftAdmissionPlugins = options.AdmissionConfig.PluginOrderOverride
+		hasSeparateKubeAdmissionChain = true
+	}
+
+	if options.KubernetesMasterConfig != nil && !hasSeparateKubeAdmissionChain {
+		// check for collisions between openshift and kube plugin config
+		for pluginName, kubeConfig := range options.KubernetesMasterConfig.AdmissionConfig.PluginConfig {
+			if openshiftConfig, exists := options.AdmissionConfig.PluginConfig[pluginName]; exists && !reflect.DeepEqual(kubeConfig, openshiftConfig) {
+				hasSeparateKubeAdmissionChain = true
+				break
+			}
+		}
+	}
+
+	if hasSeparateKubeAdmissionChain {
+		// build kube admission
+		var kubeAdmission admission.Interface
+		if options.KubernetesMasterConfig != nil {
+			var err error
+			kubeAdmission, err = newAdmissionChainFunc(kubeAdmissionPlugins, kubeAdmissionPluginConfigFilename, options.KubernetesMasterConfig.AdmissionConfig.PluginConfig, options, kubeClientSet, pluginInitializer)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+
+		// build openshift admission
+		openshiftAdmission, err := newAdmissionChainFunc(openshiftAdmissionPlugins, "", options.AdmissionConfig.PluginConfig, options, kubeClientSet, pluginInitializer)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return openshiftAdmission, kubeAdmission, nil
+	}
+
+	// if we have a unified chain, build the combined config
+	pluginConfig := map[string]configapi.AdmissionPluginConfig{}
+	if options.KubernetesMasterConfig != nil {
+		for pluginName, config := range options.KubernetesMasterConfig.AdmissionConfig.PluginConfig {
+			pluginConfig[pluginName] = config
+		}
+	}
+	for pluginName, config := range options.AdmissionConfig.PluginConfig {
+		pluginConfig[pluginName] = config
+	}
+
+	admissionChain, err := newAdmissionChainFunc(combinedAdmissionControlPlugins, "", pluginConfig, options, kubeClientSet, pluginInitializer)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return admissionChain, admissionChain, err
+}
+
+// newAdmissionChainFunc is for unit testing only.  You should NEVER OVERRIDE THIS outside of a unit test.
+var newAdmissionChainFunc = newAdmissionChain
+
+func newAdmissionChain(pluginNames []string, admissionConfigFilename string, pluginConfig map[string]configapi.AdmissionPluginConfig, options configapi.MasterConfig, kubeClientSet *internalclientset.Clientset, pluginInitializer oadmission.PluginInitializer) (admission.Interface, error) {
+	plugins := []admission.Interface{}
+	for _, pluginName := range pluginNames {
+		switch pluginName {
+		case lifecycle.PluginName:
```

Comment:
> I would assume we pass exactly the data they need (immortal namespaces, reject/admit CIDRs, LimitSecretReferences, etc). This is what we do with the other plugins, no?

We've tried to keep the interfaces generally useful (think passing client.Interface, not individual namespacers).  We could try to build a synthetic `AdmissionPluginConfig` for them, which would actually fit quite well.